tests/msc2716_test.go

```diff
@@ -0,0 +1,643 @@
+
+	timeBetweenMessagesMS := uint64(TimeBetweenMessages / time.Millisecond)
+
+	evs := make([]map[string]interface{}, count)
+	for i := 0; i < len(evs); i++ {
+		newEvent := map[string]interface{}{
+			"type":             "m.room.message",
+			"sender":           virtualUserID,
+			"origin_server_ts": insertOriginServerTs + (timeBetweenMessagesMS * uint64(i)),
+			"content": map[string]interface{}{
+				"msgtype":          "m.text",
+				"body":             fmt.Sprintf("Historical %d (chunk=%d)", i, chunkCount),
+				MSC2716_HISTORICAL: true,
+			},
+		}
+
+		evs[i] = newEvent
+	}
+
+	joinEvent := map[string]interface{}{
+		"type":             "m.room.member",
+		"sender":           virtualUserID,
+		"origin_server_ts": insertOriginServerTs,
+		"content": map[string]interface{}{
+			"membership": "join",
+		},
+		"state_key": virtualUserID,
+	}
+
+	query := make(url.Values, 2)
+	query.Add("prev_event", insertAfterEventId)
+	query.Add("user_id", virtualUserID)
+	// If provided, connect the chunk to the last insertion point
+	if chunkID != "" {
+		query.Add("chunk_id", chunkID)
+	}
+
+	b, err := json.Marshal(map[string]interface{}{
+		"events":                evs,
+		"state_events_at_start": []map[string]interface{}{joinEvent},
+	})
+	if err != nil {
+		t.Fatalf("msc2716.backfillBatchHistoricalMessages failed to marshal JSON body: %s", err)
+	}
+
+	res = c.DoFunc(
+		t,
+		"POST",
+		[]string{"_matrix", "client", "unstable", "org.matrix.msc2716", "rooms", roomID, "batch_send"},
+		client.WithRawBody(b),
+		client.WithContentType("application/json"),
+		client.WithQueries(query),
+	)
+	// Save the body so we can re-create after the buffer closes
+	body := client.ParseJSON(t, res)
+	// Since the original body can only be read once, create a new one from the body bytes we just read
+	res.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+	must.MatchResponse(t, res, match.HTTPResponse{
+		StatusCode: status,
+	})
+	// After using up the body in the must.MatchResponse above, create the body again
+	// Since the original body can only be read once, create a new one from the body bytes we just read
+	res.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+
+	chunkCount++
+
+	return res
+}
+
+func getEventsFromBatchSendResponse(t *testing.T, res *http.Response) (stateEventsIDs []string, eventIDs []string) {
+	body := client.ParseJSON(t, res)
+	// Since the original body can only be read once, create a new one from the body bytes we just read
+	res.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+
+	stateEventsIDs = client.GetJSONFieldArray(t, body, "state_events")
+	eventIDs = client.GetJSONFieldArray(t, body, "events")
+
+	return stateEventsIDs, eventIDs
+}
+
+func getNextChunkIdFromBatchSendResponse(t *testing.T, res *http.Response) (nextChunkID string) {
+	body := client.ParseJSON(t, res)
+	// Since the original body can only be read once, create a new one from the body bytes we just read
+	res.Body = ioutil.NopCloser(bytes.NewBuffer(body))
```

Comment:
We do this in a few places, like `getEventsFromBatchSendResponse` and it feels wrong that you need to keep doing this. Why aren't you just reading the response body _once_ and using the resulting `body []byte` in these functions..? We're doing needless processing (parsing JSON) and copies currently.

Reply:
It feels weird to have some functions accept responses, and some accept body and will totally depend on the order they are called and used.

Reply:
My proposal would be to perform the HTTP request and check the status code immediately, then use `body []byte` everywhere since that's all you actually care about. The constant re-assignments to `res.Body` is needless boilerplate.